plugins/GranularPitchShifter/GranularPitchShifterEffect.h

```diff
@@ -0,0 +1,168 @@
+
+#include "Effect.h"
+#include "GranularPitchShifterControls.h"
+
+#include "BasicFilters.h"
+#include "interpolation.h"
+
+namespace lmms
+{
+
+constexpr float PrefilterBandwidth = 0.96f;// 96% of nyquist
+constexpr double GlideSnagRadius = 0.001;
+constexpr int SafetyLatency = 3;
+constexpr float RangeSeconds[5] = {5, 10, 40, 40, 120};
+
+
+class GranularPitchShifterEffect : public Effect
+{
+public:
+	GranularPitchShifterEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key);
+	~GranularPitchShifterEffect() override = default;
+	bool processAudioBuffer(sampleFrame* buf, const fpp_t frames) override;
+
+	EffectControls* controls() override
+	{
+		return &m_granularpitchshifterControls;
+	}
+	
+	// double index and fraction are required for good quality
+	float getHermiteSample(double index, int ch) {
+		const int index_floor = static_cast<int>(index);
+		const double fraction = index - index_floor;
+		
+		float v0, v1, v2, v3;
+
+		if (index_floor == 0) {
+			v0 = m_ringBuf[m_ringBuf.size() - 1][ch];
+		} else {
+			v0 = m_ringBuf[index_floor - 1][ch];
+		}
+		
+		v1 = m_ringBuf[index_floor][ch];
+		
+		if(index_floor >= m_ringBuf.size() - 2) {
+			v2 = m_ringBuf[(index_floor + 1) % m_ringBuf.size()][ch];
+			v3 = m_ringBuf[(index_floor + 2) % m_ringBuf.size()][ch];
+		} else {
+			v2 = m_ringBuf[index_floor + 1][ch];
+			v3 = m_ringBuf[index_floor + 2][ch];
+		}
+		
+		return hermiteInterpolate(v0, v1, v2, v3, static_cast<float>(fraction));
+	}
+	
+	// adapted from signalsmith's crossfade approximation:
+	// https://signalsmith-audio.co.uk/writing/2021/cheap-energy-crossfade
+	float cosHalfWindowApprox(float x, float k) {
+		float A = x * (1 - x);
+		float B = A * (1 + k * A);
+		float C = (B + x);
+		return C * C;
+	}
+	// 1-2 fades between equal-gain and equal-power
+	float cosWindowApproxK(float p) {
+		return -6.0026608f + p * (6.8773512f - 1.5838104f * p);
+	}
+	
+	void sampleRateNeedsUpdate() { m_sampleRateNeedsUpdate = true; }
+	
+	void changeSampleRate();
+
+private:
+	struct PrefilterLowpass
+	{
+		float m_v0z = 0.0f, m_v1 = 0.0f, m_v2 = 0.0f;
+		float m_g1, m_g2, m_g3, m_g4;
+
+		void setCoefs(float sampleRate, float cutoff)
+		{
+		    const float g = std::tan(F_PI * cutoff / sampleRate);
+		    const float ginv = g / (1.0f + g * (g + 1.414213562));
+		    m_g1 = ginv;
+		    m_g2 = 2.0f * (g + 1.414213562) * ginv;
+		    m_g3 = g * ginv;
+		    m_g4 = 2.0f * ginv;
+		}
+
+		float process(float input)
+		{
+		    const float v1z = m_v1;
+		    const float v3 = input + m_v0z - 2.0f * m_v2;
+		    m_v1 += m_g1 * v3 - m_g2 * v1z;
+		    m_v2 += m_g3 * v3 + m_g4 * v1z;
+		    m_v0z = input;
+		    return m_v2;
+		}
+	};
+
+	struct Grain
+	{
+		Grain(double grainSpeedL, double grainSpeedR, double phaseSpeedL, double phaseSpeedR, double readPointL, double readPointR) :
+			readPoint{readPointL, readPointR},
+			phaseSpeed{phaseSpeedL, phaseSpeedR},
+			grainSpeed{grainSpeedL, grainSpeedR},
+			phase{0} {}
```

Comment:
A little hard to see, just make it more obvious.
```suggestion
			phase{0}
		{}
```